### PR TITLE
135 enable Gather Coverage Data

### LINKED ISSUE #135: Enable "Gather coverage data" in example app and pod framework
I was asked to submit a PR by @segiddins [here](https://github.com/CocoaPods/CocoaPods/issues/3706#issuecomment-156587575). Any reason not to just enable it for the default project through `pod lib create`? It could be made an option, but I don't know of any downsides to just enabling coverage all the time, and making the pod a build dependency of the example project.

### PR DESCRIPTION
**Pull request by ZevEisenberg**
Enabled **Gather coverage data** in example project. Going to dig into `xcodeproj` to figure out how to enable it for the pod framework, and make the pod framework a build target of the example project, but I wanted to post this early just to make sure I'm going about it the right way. Is that something I would change in `ProjectManipulator.rb`, or does it have to be part of CocoaPods proper?

Fixes #135.

### CODE REVIEW

**Comment by segiddins** [comment]
:+1: 
I don't think a change to Xcodeproj is warranted, as Xcodeproj by default generates things in line with Xcode defaults -- its up to the consumers of it to make changes as desired.

**Comment by orta** [comment]
This came up in a discussion earlier, think current best course of action is adding the glad in a post install hook in the podfile

**Comment by ZevEisenberg** [comment]
In that case, is anything else required to merge this PR? And can you point me to a resource for the changes you mentioned for the pod file? I don't know what a "glad" is, unless that's typo. (Even if it is, I still don't know what you mean.)

**Comment by ZevEisenberg** [comment]
Looked at it bit more. Post-install hook in the example project's Podfile definitely looks like the way to go. So hold off on merging this PR, because I'll add the post-install script once I've figured out The Things. cc @kingofbrian who was looking at this with me.

**Comment by ZevEisenberg** [comment]
Been working on this. Ran into two sticking points. Anyone know the answers to either of these Stack Overflow questions that I posted?

1. [Get path to particular scheme](http://stackoverflow.com/questions/33848281/get-path-to-particular-scheme)
1. [Modify XCScheme and write it back to disk in `post_install` script](http://stackoverflow.com/questions/33848302/modify-xcscheme-and-write-it-back-to-disk-in-post-install-script)

**Comment by ZevEisenberg** [comment]
Oh, I get it. "Glad" is a typo for "flag". I are smart programer.

**Comment by ZevEisenberg** [comment]
Updated PR with a `post_install` script. i-have-no-idea-what-im-doing.jpeg. I also wasn't exactly sure how to test it, but @orta or @segiddins, if one of you could let me know, I'll add a section to the readme for future people.

**Comment by ZevEisenberg** [comment]
Hold off on merging. Script doesn't check for existing build action, so just keeps adding it. Also doesn't seem to be adding it quite correctly. Xcode uses double quotes, but Xcodeproj is using single quotes. Not sure why. Looking into it.

File: templates/ios/Example/Podfile [hunk 1]
```diff
@@ -10,3 +10,27 @@ target '${POD_NAME}_Tests', :exclusive => true do
...
   ${INCLUDED_PODS}
 end
+
+# Add ${POD_NAME} target to build step of the example project
+# so that we get code coverage reports for ${POD_NAME}
+post_install do |installer|
+  # get ${POD_NAME} target from Pods project
+  pods_project = installer.pods_project
+  pods_targets = pods_project.targets
+  pod_target = pods_targets.select { |target| target.name == "${POD_NAME}" }.uniq.first
```
Comment (segiddins) [inline templates/ios/Example/Podfile]: this could just be `find`

File: templates/ios/Example/Podfile [hunk 2]
```diff
@@ -10,3 +10,27 @@ target '${POD_NAME}_Tests', :exclusive => true do
...
+
+  # get example project scheme
+  example_project = installer.aggregate_targets.map(&:user_project_path).uniq.first
```
Comment (segiddins) [inline templates/ios/Example/Podfile]: no need to `uniq`

**Comment by orta** [comment]
OK, hrm, I wish we had a better way of doing this. I'm not too comfortable with all that extra code in the Podfile. 

There's obviously value in doing coverage data, I wonder if this should be a better feature in CocoaPods or as a plugin,

**Comment by ZevEisenberg** [comment]
I agree about the podfile. Happy to hear suggestions. I can take a crack at implementing something, but I don't know anything about how CocoaPods does its magic, so I can't intelligently suggest anything specific.

File: templates/ios/Example/Podfile (hunk 1 above)
Comment (ZevEisenberg) [inline templates/ios/Example/Podfile]: Thanks. This is my first non-Podfile Ruby code. i-have-no-idea-what-im-doing.jpg

**Comment by orta** [comment]
Could the code be ran after the first Pod install from the `pod lib create` installer? It could be possible that shared scheme stay between `pod install`

**Comment by ZevEisenberg** [comment]
@orta that may work - I forget if `pod install` currently blows away the change that is required. The current problem I'm working on is that the Pod target must be added as a build action dependency of the example project scheme, and I couldn't figure out a way to add that in the template, and I'm having a heck of a time getting it working in a post-install script.

**Comment by AvdLee** [comment]
@ZevEisenberg I'm curious about the current status here. I'm looking for the same solution for my CocoaPods, but including your post script inside my podfile doesn't seem to do the trick.

Can you get me up and running with this?

**Comment by ZevEisenberg** [comment]
@AvdLee I've never been able to get it to work. This PR doesn't fully do it - there are other steps needed, and I haven't been able to look at it in a while.

**Comment by fulldecent** [comment]
@ZevEisenberg As just a small step, could you pleas resolve merge conflicts and maybe others can pick up after that

**Comment by ZevEisenberg** [comment]
@fulldecent I rebased my fork on top of latest master. Should be good to go.

**Comment by orta** [comment]
Hrm, yeah, I also couldn't get this to work after playing with it locally, even when defining your own scheme just for the framework, coverage is still applied on an app level

screen shot 2016-05-02 at 07 17 22

**Comment by orta** [comment]
I'm not sure if it's possible to have hosted framework tests that use code-coverage, so this option may only e available if you choose to not have an example application perhaps.

**Comment by ZevEisenberg** [comment]
I chatted with @orta and @jessesquires at WWDC about this. One approach that came up (thanks, @jessesquires) was that instead of modifying the setting in the scheme, you can instruct xcodebuild to gather coverage data while testing via a command line argument. In terms of this bug, we could potentially modify the default .travis.yml in the pod template to enable coverage gathering on Travis by default. It wouldn't solve the problem for everyone, but since Travis is the de facto CI system for CocoaPods (as evidenced by the presence of .travis.yml in this template), it could cover (haha) the common case. If you're using a difference CI system, you can always port the xcodebuild parameter to your own build scripts.

**Comment by orta** [comment]
I'm good with that 👍

**Comment by fulldecent** [comment]
Competing PR at https://github.com/CocoaPods/pod-template/pull/181

**Comment by dnkoutso** [comment]
This is too old of PR and can possibly now be done via the `scheme` DSL which can easily be extended to support code coverage. Then this tool can query the user if code coverage should be enabled or not.

Closing as too old for now.

**Pull request closed without merging**